contracts/dual_token/DualStakingToken.sol

```diff
@@ -271,22 +304,32 @@ contract DualStakingToken {
         IService(serviceRegistry).approve(stakingInstance, serviceId);
         // Stake service
         IStaking(stakingInstance).stake(serviceId);
+
+        _locked = 1;
     }
 
     /// @dev Unstakes OLAS service Id and unbonds staking token amount.
     /// @param serviceId OLAS driven service Id.
     function unstake(uint256 serviceId) external {
-        StakerInfo storage stakerInfo = mapStakerStakerInfos[serviceId];
+        // Reentrancy guard
+        if (_locked > 1) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        StakerInfo storage stakerInfo = mapStakerInfos[serviceId];
         // Check for staker existence
         if (stakerInfo.account == address(0)) {
             revert();
         }
 
-        // Get staking instance
-        address stakingInstance = stakerInfo.stakingInstance;
+        // Get staked service state
+        IStaking.StakingState stakingState = IStaking(stakingInstance).getStakingState(serviceId);
 
-        // Perform checkpoint first as there might be more rewards
-        checkpoint(stakingInstance);
+        // Perform checkpoint first as there might be more rewards, only if the service is still staked
+        if (stakingState == IStaking.StakingState.Staked) {
+            checkpoint();
+        }
```

Comment:
No need to call checkpoint if the service is not staked - no incentive for the unstaked service Id.